contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -0,0 +1,903 @@
+}
+
+void Filter::onDestroy() {
+  ENVOY_LOG(info, "golang filter on destroy");
+
+  {
+    Thread::LockGuard lock(mutex_);
+    if (has_destroyed_) {
+      ENVOY_LOG(warn, "golang filter has been destroyed");
+      return;
+    }
+    has_destroyed_ = true;
+  }
+
+  if (dynamic_lib_ == nullptr) {
+    ENVOY_LOG(error, "golang filter dynamicLib is nullPtr.");
+    return;
+  }
+
+  try {
+    ASSERT(req_ != nullptr);
+    auto& state = getProcessorState();
+    auto reason = state.isProcessingInGo() ? DestroyReason::Terminate : DestroyReason::Normal;
+
+    dynamic_lib_->envoyGoFilterOnHttpDestroy(req_, int(reason));
+  } catch (...) {
+    ENVOY_LOG(error, "golang filter onDestroy do destoryStream catch "
+                     "unknown exception.");
+  }
+}
+
+void Filter::onStreamComplete() {
+  addGolangMetadata("cost_decode", cost_time_decode_);
+  addGolangMetadata("cost_encode", cost_time_encode_);
+  addGolangMetadata("cost_total", cost_time_decode_ + cost_time_encode_ + cost_time_mem_);
+}
+
+void Filter::addGolangMetadata(const std::string& k, const uint64_t v) {
+  ProtobufWkt::Struct value;
+  ProtobufWkt::Value val;
+  val.set_number_value(static_cast<double>(v));
+  (*value.mutable_fields())[k] = val;
+  decoding_state_.getFilterCallbacks()->streamInfo().setDynamicMetadata("golang.extension", value);
+}
+
+// access_log is executed before the log of the stream filter
+void Filter::log(const Http::RequestHeaderMap*, const Http::ResponseHeaderMap*,
+                 const Http::ResponseTrailerMap*, const StreamInfo::StreamInfo&) {
+  // Todo log phase of stream filter
+}
+
+/*** common APIs for filter, both decode and encode ***/
+
+GolangStatus Filter::doHeadersGo(ProcessorState& state, Http::RequestOrResponseHeaderMap& headers,
+                                 bool end_stream) {
+  ENVOY_LOG(debug, "golang filter passing data to golang, state: {}, phase: {}, end_stream: {}",
+            state.stateStr(), state.phaseStr(), end_stream);
+
+  try {
+    if (req_ == nullptr) {
+      // req is used by go, so need to use raw memory and then it is safe to release at the gc
+      // finalize phase of the go object.
+      req_ = new httpRequestInternal(weak_from_this());
+      req_->configId = getMergedConfigId(state);
+      req_->plugin_name.data = config_->pluginName().data();
+      req_->plugin_name.len = config_->pluginName().length();
+    }
+
+    req_->phase = static_cast<int>(state.phase());
+    {
+      Thread::LockGuard lock(mutex_);
+      headers_ = &headers;
+    }
+    auto status = dynamic_lib_->envoyGoFilterOnHttpHeader(req_, end_stream ? 1 : 0, headers.size(),
+                                                          headers.byteSize());
+    return static_cast<GolangStatus>(status);
+
+  } catch (const EnvoyException& e) {
+    ENVOY_LOG(error, "golang filter doHeadersGo catch: {}.", e.what());
+
+  } catch (...) {
+    ENVOY_LOG(error, "golang filter doHeadersGo catch unknown exception.");
+  }
```

Comment:
Please remove try/catch handling from data plane code. Why/how can this code throw? Also please never catch all exceptions. Let's just remove exception handling entirely from all of this.

If somehow go panics are translated to exceptions, let's have the go code catch/translate to error returns.

Reply:
Done. Thanks.